File: android/revenuedog/src/main/kotlin/com/revenuedog/purchases/PurchasesError.kt
package com.revenuedog.purchases

/**
 * Stable error codes thrown by every RevenueDog SDK operation, shared verbatim with the iOS SDK.
 */
enum class PurchasesErrorCode(val description: String) {
    /** The backend could not be reached (timeout, DNS, connection failure). */
    NETWORK_ERROR("A network error occurred."),

    /** The user cancelled the purchase flow. */
    PURCHASE_CANCELLED("The purchase was cancelled by the user."),

    /** The purchase request was invalid (e.g. item already owned, malformed request). */
    PURCHASE_INVALID("The purchase request was invalid."),

    /** The requested product could not be found on the store. */
    PRODUCT_NOT_FOUND("The product was not found on the store."),

    /** The backend's store validator rejected the receipt. */
    RECEIPT_VALIDATION_FAILED("Receipt validation failed."),

    /** Google Play Billing reported a problem. */
    STORE_PROBLEM("There was a problem with the Play Store."),

    /** The SDK is misconfigured (bad API key, missing base URL, logOut while anonymous, ...). */
    CONFIGURATION_ERROR("The SDK is not configured correctly."),

    /** The purchase is pending external action (e.g. cash payment); no receipt was posted. */
    PENDING("The purchase is pending and not yet complete."),

    /** Anything else. */
    UNKNOWN("An unknown error occurred.")
}

/**
 * The single error type thrown by all RevenueDog APIs.
 *
 * @property code stable, programmatically matchable error code.
 * @property underlyingErrorMessage detail from the backend or Play Billing, when available.
 */
class PurchasesError(
    val code: PurchasesErrorCode,
    val underlyingErrorMessage: String? = null,
    cause: Throwable? = null
) : Exception(underlyingErrorMessage ?: code.description, cause) {

    override fun toString(): String =
        "PurchasesError(code=$code, message=${underlyingErrorMessage ?: code.description})"
}

File: ios/Sources/RevenueDog/PurchasesError.swift
import Foundation

/// The single error type thrown by every throwing ``Purchases`` method.
///
/// Inspect ``PurchasesError/code`` for stable, switchable error handling. The
/// ``PurchasesError/underlyingError`` carries the originating error (a
/// `URLError`, StoreKit error, decoding error, …) when one exists.
public struct PurchasesError: Error, CustomStringConvertible {

    /// Stable, machine-readable classification of the failure.
    ///
    /// These cases are part of the public contract and match the Android SDK.
    public enum Code: String, Sendable, Equatable {
        /// A network request failed (offline, timeout, transport error, 5xx).
        case networkError
        /// The user cancelled the native purchase sheet.
        case purchaseCancelled
        /// The purchase could not be completed or verified by the store.
        case purchaseInvalid
        /// A requested product or package could not be found.
        case productNotFound
        /// The backend rejected the submitted receipt (HTTP 422).
        case receiptValidationFailed
        /// A generic problem talking to the underlying store.
        case storeProblem
        /// The SDK was used incorrectly (e.g. not configured, missing key).
        case configurationError
        /// The purchase is pending external action (deferred / ask-to-buy).
        case pending
        /// An unclassified error.
        case unknown
    }

    /// The stable error classification.
    public let code: Code

    /// Human-readable description of what went wrong.
    public let message: String

    /// The originating error, when one exists.
    public let underlyingError: Error?

    public init(code: Code, message: String, underlyingError: Error? = nil) {
        self.code = code
        self.message = message
        self.underlyingError = underlyingError
    }

    public var description: String {
        var text = "PurchasesError(code: \(code.rawValue), message: \(message)"
        if let underlyingError {
            text += ", underlying: \(underlyingError)"
        }
        text += ")"
        return text
    }

    // MARK: - Convenience factories

    static func network(_ message: String, _ underlying: Error? = nil) -> PurchasesError {
        PurchasesError(code: .networkError, message: message, underlyingError: underlying)
    }

    static func configuration(_ message: String) -> PurchasesError {
        PurchasesError(code: .configurationError, message: message)
    }

    static func store(_ message: String, _ underlying: Error? = nil) -> PurchasesError {
        PurchasesError(code: .storeProblem, message: message, underlyingError: underlying)
    }

    static func cancelled(_ message: String = "The purchase was cancelled.") -> PurchasesError {
        PurchasesError(code: .purchaseCancelled, message: message)
    }

    static func productNotFound(_ message: String) -> PurchasesError {
        PurchasesError(code: .productNotFound, message: message)
    }

    static func receiptValidationFailed(_ message: String) -> PurchasesError {
        PurchasesError(code: .receiptValidationFailed, message: message)
    }

    static let pendingPurchase = PurchasesError(
        code: .pending,
        message: "The purchase is pending external confirmation."
    )

    static func unknown(_ message: String, _ underlying: Error? = nil) -> PurchasesError {
        PurchasesError(code: .unknown, message: message, underlyingError: underlying)
    }
}
